### ISSUE TITLE
Bug | Uncaught SyntaxError: Cannot use import statement outside a module

### ISSUE THREAD

**Issue by KleinDevDE** [orestbida/cookieconsent#410]
**Cookie-Consent version:** 2.8.8
**Tested with:** 2.8.6, 2.8.8
**Affected browsers:** `Chrome v107.0.5304.88 (64-Bit)`, `Edge v107.0.1418.26 (64-Bit)`, `Firefox v106.0.3 (64-Bit)`

There is a bug that seems to affect all common browsers.
Chrome, Edge and Firefox have been tested.

In Chromium based browsers, the small window with the Accept and Deny buttons appears in the bottom right corner. (Image 1)
As soon as you press accept or reject here, the error message listed below appears. 
Actions like activating previously blocked scripts (e.g. because of targeting) work fine.

In Firefox, however, this window does not appear at all, the error message below appears immediately.

Chromium and Firefox browsers can both display the settings window (image2).
I can also accept or reject everything via it, subsequently the scripts are also loaded or blocked, as expected.

Normally this would then be considered a low priority bug.
But since in Firefox the one window on the bottom right (image 1) does not appear at all, it is, in my estimation, a bug with a higher priority.

Image and source code of cookieconsent_init.js is below, partly in the expandable spoiler to make the ticket not quite too big.

---

**The way its included**
```
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v2.8.8/dist/cookieconsent.css" media="print" onload="this.media='all'">
<script defer src="https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v2.8.8/dist/cookieconsent.js"></script>
<script defer src="{{asset('js/cookieconsent_init.js')}}"></script>
```

**On Chrome v107.0.5304.88 (64-Bit)**
The Modal (image1) will be displayed on the bottom right corner.
After clicking on accept or decline, this error will come up.
```
VM30:2 Uncaught SyntaxError: Cannot use import statement outside a module
    at o (cookieconsent.js:7:13770)
    at ln (cookieconsent.js:7:13808)
    at cookieconsent.js:7:16734
    at U.accept (cookieconsent.js:7:16945)
    at HTMLButtonElement.<anonymous> (cookieconsent.js:7:3851)
```

**On Edge v107.0.1418.26 (64-Bit)**
The Modal (image1) will be displayed on the bottom right corner.
After clicking on accept or decline, this error will come up.

```
       Uncaught SyntaxError: Cannot use import statement outside a module
    at o (cookieconsent.js:7:13770)
    at ln (cookieconsent.js:7:13808)
    at cookieconsent.js:7:16734
    at U.accept (cookieconsent.js:7:16945)
    at HTMLButtonElement.<anonymous> (cookieconsent.js:7:3851)
```

**On Firefox v106.0.3 (64-Bit)**
The modal (image1) will not be displayed, the error comes right after the loading of the site finished.
```
Uncaught SyntaxError: import declarations may only appear at top level of a module
    o https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v2.8.8/dist/cookieconsent.js:7
    ln https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v2.8.8/dist/cookieconsent.js:7
    run https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v2.8.8/dist/cookieconsent.js:7
    <anonymous> https://kolibri.dev.kompadu.de/js/cookieconsent_init.js:3
```

---

**Image 1**
![grafik](https://user-images.githubusercontent.com/13904220/199482338-9390d9cb-4587-466c-aafd-383608923117.png)

**image 2**
![grafik](https://user-images.githubusercontent.com/13904220/199485122-98dc8e87-5e04-48a0-9fdb-09968582511d.png)

<details>
    <summary>cookieconsent_init.js</summary>

```
window.cc = initCookieConsent(); // run plugin with your configuration

cc.run({
    current_lang: 'en',
    autoclear_cookies: true,
    // default: false
    page_scripts: true,
    // default: false
    // mode: 'opt-in'                          // default: 'opt-in'; value: 'opt-in' or 'opt-out'
    // delay: 0,                               // default: 0
    // auto_language: null                     // default: null; could also be 'browser' or 'document'
    // autorun: true,                          // default: true
    // force_consent: false,                   // default: false
    // hide_from_bots: false,                  // default: false
    // remove_cookie_tables: false             // default: false
    // cookie_name: 'cc_cookie',               // default: 'cc_cookie'
    // cookie_expiration: 182,                 // default: 182 (days)
    // cookie_necessary_only_expiration: 182   // default: disabled
    // cookie_domain: location.hostname,       // default: current domain
    // cookie_path: '/',                       // default: root
    // cookie_same_site: 'Lax',                // default: 'Lax'
    // use_rfc_cookie: false,                  // default: false
    // revision: 0,                            // default: 0
    onFirstAction: function onFirstAction(user_preferences, cookie) {// callback triggered only once
    },
    onAccept: function onAccept(cookie) {// ...
    },
    onChange: function onChange(cookie, changed_preferences) {// ...
    },
    languages: {
        'en': {
            consent_modal: {
                title: 'Wir nutzen Cookies!',
                description: 'Hi, diese Webseite nutzt essentielle Cookies um eine reibungslose Erfahrung zu gewährleisten und Trackingcookies um besser zu verstehen wie du mit der Webseite interargierst. Die Trackingcookies sind nur aktiv nachdem du zugestimmt hast. <button type="button" data-cc="c-settings" class="cc-link">Konfigurieren</button>.',
                primary_btn: {
                    text: 'Alle akzeptieren',
                    role: 'accept_all' // 'accept_selected' or 'accept_all'
                },
                secondary_btn: {
                    text: 'Alle ablehnen',
                    role: 'accept_necessary' // 'settings' or 'accept_necessary'
                }
            },
            settings_modal: {
                title: 'Cookie preferences',
                save_settings_btn: 'Einstellungen speichern',
                accept_all_btn: 'Alle akzeptieren',
                reject_all_btn: 'Alle ablehnen',
                close_btn_label: 'Schließen',
                cookie_table_headers: [
                    {col1: 'Name'},
                    {col2: 'Domain'},
                    {col3: 'Verfallsdatum'},
                    {col4: 'Beschreibung'}
                ],
                blocks: [
                    {
                        title: 'Cookie Verwendung 📢',
                        description: 'Wir verwenden Cookies, um die grundlegenden Funktionen der Website zu gewährleisten und um Ihr Online-Erlebnis zu verbessern. Sie können für jede Kategorie wählen, ob Sie sich ein- oder austragen möchten. Für weitere Einzelheiten zu Cookies und anderen sensiblen Daten lesen Sie bitte die vollständige <a href="/datenschutz" class="cc-link">Datenschutzerklärung</a>.'
                    }, {
                        title: 'Streng notwendige Cookies',
                        description: 'Diese Cookies sind für das ordnungsgemäße Funktionieren der Website unerlässlich. Ohne diese Cookies würde die Website nicht richtig funktionieren.',
                        toggle: {
                            value: 'necessary',
                            enabled: true,
                            readonly: true // cookie categories with readonly=true are all treated as "necessary cookies"

                        }
                    },
                    {
                        title: 'Leistungs- und Analyse-Cookies',
                        description: 'Diese Cookies ermöglichen es der Website, sich an die von Ihnen in der Vergangenheit getroffenen Auswahlen zu erinnern. Meistens werden nicht alle in dieser Auflistung genannten Cookies verwendet.',
                        toggle: {
                            value: 'analytics',
                            // your cookie category
                            enabled: false,
                            readonly: false
                        },
                        cookie_table: [// list of all expected cookies
                            {
                                col1: '^_ga',
                                // match all cookies starting with "_ga"
                                col2: 'google.com',
                                col3: '2 Jahre',
                                col4: 'Identifizieren des Benutzers und speichern von Daten',
                                is_regex: true
                            },
                            {
                                col1: '_gid',
                                col2: 'google.com',
                                col3: '1 Tag',
                                col4: 'Identifizieren des Benutzers'
                            },
                            {
                                col1: '_gat',
                                col2: 'google.com',
                                col3: '1 Minute',
                                col4: 'Limitierung der Anfragen'
                            },
                            {
                                col1: '^_gac_',
                                col2: 'google.com',
                                col3: '90 Tage',
                                col4: 'Beinhaltet Kampagnen-bezogene Informationen.',
                                is_regex: true
                            },
                            {
                                col1: '_gat',
                                col2: 'google.com',
                                col3: '1 Minute',
                                col4: 'Limitierung der Anfragen'
                            },
                        ]
                    },
                    {
                        title: 'Cookies für Werbung und Zielgruppenansprache',
                        description: 'Diese Cookies sammeln Informationen darüber, wie Sie die Website nutzen, welche Seiten Sie besucht und welche Links Sie angeklickt haben. Alle Daten sind anonymisiert und können nicht verwendet werden, um Sie zu identifizieren.',
                        toggle: {
                            value: 'targeting',
                            enabled: false,
                            readonly: false
                        }
                    },
                    {
                        title: 'Verbindung zu externen Diensten (Captcha)',
                        description: 'Für das Kontaktformular verwenden wir das Google reCaptcha, um Spam entgegen zu wirken. Hierzu wird eine Verbindung zwischen Ihnen und den amerikanischen Servern hergestellt, welche hierdurch zB Ihre IP-Adresse in erfahrung bringen können. Weitere Informationen finden Sie in der <a href="/datenschutz" class="cc-link">Datenschutzerklärung</a>.',
                        toggle: {
                            value: 'captcha',
                            enabled: false,
                            readonly: false
                        }
                    },
                    {
                        title: 'Verbindung zu externen Diensten (Maps)',
                        description: 'Wir verwenden zB die Google-Maps integration zur leichteren Darstellung diverser Standorte. Hierzu wird eine Verbindung zwischen Ihnen und den amerikanischen Servern hergestellt, welche hierdurch zB Ihre IP-Adresse in erfahrung bringen können. Weitere Informationen finden Sie in der <a href="/datenschutz" class="cc-link">Datenschutzerklärung</a>.',
                        toggle: {
                            value: 'maps',
                            enabled: false,
                            readonly: false
                        }
                    },
                    {
                        title: 'Weitere Informationen',
                        description: 'Wenn Sie Fragen zu unserer Cookie-Politik und Ihren Wahlmöglichkeiten haben, wenden Sie sich bitte an <a class="cc-link" href="/kontakt">Kontakt</a>.'
                    }
                ]
            }
        }
    }
});
```

</details>

**Comment by orestbida** [comment]
Hi @KleinDevDE, I cannot reproduce this issue. There are no `import` statements in the plugin, so this is very likely specific to your project.

Here is fully working [demo with your config](https://stackblitz.com/edit/web-platform-t1yyrs?file=cookieconsent-init.js).

**Comment by KleinDevDE** [comment]
The cause has now finally been found.

It is because one of the "necessary" scripts uses the syntax of "import X from Y".
So, when CookieConsent "activated" this script tag, the error message occurred.
So the browser thinks it's CC because it was loaded and executed by CC.

This script was initialized as follows:
```
<script type="module" data-cookiecategory="necessary">
        import Swiper from 'https://cdn.jsdelivr.net/npm/swiper@8/swiper-bundle.esm.browser.min.js'
...
</script>
```

@orestbida , would it be possible that CC by itself gives an info where error XY was throwned?
This avoids irritations like mine here, and gives direct information that it is not the CC but one of the scripts that were activated.

**Comment by orestbida** [comment]
Mhm, correct me if I'm wrong, but this is a runtime error caused by another script. I'm not aware of any way to detect if another script threw an exception or not.

For the sake of clarity: the plugin does not handle `type=module` script tags. The `type` attribute is completely removed when the category is enabled. Also, a script with an invalid javascript mime type is required — e.g. `type="text/plain"` — to prevent its execution.

**Comment by zangab** [comment]
the issue is, that the consent is changing the type from `type="text/plain"` to `type="text/javascript"`. however, what we would need to make this work is a some kind of config to change it to `type="module"`. this is the actual problem here.

**Comment by zangab** [comment]
ok, seems like this is already possible by setting the `data-type` attribute. it's missing in the docs though?


```js
// from source code
curr_script.type = curr_script.getAttribute('data-type') || 'text/javascript';
```
